### PR TITLE
Use gnu gold or clang lld linkers if available

### PR DESCRIPTION
**Pull request by seelabs**
These linkers significantly improve link times

### CODE REVIEW

File: Builds/CMake/CMakeFuncs.cmake [hunk 1]
```diff
@@ -482,6 +482,15 @@ macro(setup_build_boilerplate)
 
   if (is_gcc)
     add_compile_options(-Wno-unused-but-set-variable -Wno-deprecated)
+
+    # use gold linker if available
+    execute_process(
+      COMMAND ${CMAKE_C_COMPILER} -fuse-ld=gold -Wl,--version
```
Comment (mellery451) [inline Builds/CMake/CMakeFuncs.cmake]: I notice here you are using the C compiler here and in scons you are using the CXX compiler. I don't think it makes any practical difference, but perhaps stick with CXX for consistency?

File: SConstruct [hunk 2]
```diff
@@ -554,6 +554,13 @@ def config_env(toolchain, variant, env):
         if toolchain == 'clang':
             env.Append(CCFLAGS=['-Wno-redeclared-class-member'])
             env.Append(CPPDEFINES=['BOOST_ASIO_HAS_STD_ARRAY'])
+            try:
+                ldd_ver = subprocess.check_output([env['CLANG_CXX'], '-fuse-ld=lld', '-Wl,--version'],
+                                            stderr=subprocess.STDOUT).strip()
```
Comment (mellery451) [inline SConstruct]: do we want to check the contents of stdout (the version string) like in CMake or is zero exit status fine?

File: Builds/CMake/CMakeFuncs.cmake [hunk 3]
```diff
@@ -524,6 +533,15 @@ macro(setup_build_boilerplate)
...
       add_definitions(-DBOOST_ASIO_HAS_STD_ARRAY)
+
+      # use ldd linker if available
+      execute_process(
+        COMMAND ${CMAKE_C_COMPILER} -fuse-ld=lld -Wl,--version
+        ERROR_QUIET OUTPUT_VARIABLE LD_VERSION)
+      if ("${LD_VERSION}" MATCHES "LLD")
+        append_flags(CMAKE_EXE_LINKER_FLAGS -fuse-ld=lld)
+      endif ()
+      unset(LD_VERSION)
```
Comment (mellery451) [inline Builds/CMake/CMakeFuncs.cmake]: this doesn't seem to work on macos (it's not fatal, but the alternative linker is not detected). I'll investigate separately if the llvm linker can be used on macos, but that should not hold up this PR.

Comment (seelabs) [inline Builds/CMake/CMakeFuncs.cmake]: After investigating with @mellery451 lld was not available on his mac.

File: SConstruct (hunk 2 above)
Comment (seelabs) [inline SConstruct]: zero exit is fine.

File: Builds/CMake/CMakeFuncs.cmake (hunk 1 above)
Comment (seelabs) [inline Builds/CMake/CMakeFuncs.cmake]: It doesn't really matter, but it's easy enough to make the change. Will do.

**Comment by seelabs** [comment]
In 0.60.0-b7

**Pull request closed without merging**